pkg/fqdn/dnsproxy/shared_client.go

```diff
@@ -0,0 +1,357 @@
+	// must be buffered to be able to send a trigger while the receive loop is not yet ready to
+	// receive the trigger, as we do not want to stall the sender when the receiver is blocking
+	// on the read operation.
+	receiverTrigger := make(chan struct{}, 1)
+	triggerReceiver := func() {
+		select {
+		case receiverTrigger <- struct{}{}:
+		default:
+		}
+	}
+
+	// Receive loop
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		defer close(responses)
+
+		// No point trying to receive until the first request has been successfully sent, so
+		// wait for a trigger first. receiverTrigger is buffered, so this is safe
+		// to do, even if the sender sends the trigger before we are ready to receive here.
+		<-receiverTrigger
+
+		for {
+			// This will block but eventually return an i/o timeout, as we always set
+			// the timeouts before sending anything
+			r, err := conn.ReadMsg()
+			if err == nil {
+				responses <- sharedClientResponse{r, 0, nil}
+				continue // receive immediately again
+			}
+
+			// handler is not reading on the channel after closing.
+			// UDP connections return net.ErrClosed, while TCP/TLS connections are read
+			// via the io package, which return io.EOF.
+			if errors.Is(err, net.ErrClosed) || errors.Is(err, io.EOF) {
+				return
+			}
+
+			// send error response to cancel all current requests.
+			responses <- sharedClientResponse{nil, 0, err}
+
+			// wait for a trigger from the handler after any errors. Re-reading in
+			// this condition could busy loop, e.g., if a read timeout occurred.
+			// receiverTrigger is buffered so that we catch the trigger that may
+			// have been sent while we sent the error response above.
+			_, ok := <-receiverTrigger
+			if !ok {
+				return // exit immediately when the trigger channel is closed
+			}
+		}
+	}()
+
+	type waiter struct {
+		ch    chan sharedClientResponse
+		start time.Time
+	}
+	waitingResponses := make(map[uint16]waiter)
+	defer func() {
+		conn.Close()
+		close(receiverTrigger)
+
+		// Drain responses send by receive loop to allow it to exit.
+		// It may be repeatedly reading after an i/o timeout, for example.
+		for range responses {
+		}
+
+		for _, waiter := range waitingResponses {
+			waiter.ch <- sharedClientResponse{nil, 0, net.ErrClosed}
+			close(waiter.ch)
+		}
+	}()
+
+	for {
+		select {
+		case req, ok := <-requests:
+			if !ok {
+				// 'requests' is closed when SharedClient is recycled, which happens
+				// responeses (or errors) have been received and there are no more
+				// requests to be sent.
+				return
+			}
+			start := time.Now()
+
+			// Check if we already have a request with the same id
+			// Due to birthday paradox and the fact that ID is uint16
+			// it's likely to happen with small number (~200) of concurrent requests
+			// which would result in goroutine leak as we would never close req.ch
+			if _, duplicate := waitingResponses[req.msg.Id]; duplicate {
+				for n := 0; n < 5; n++ {
```

Comment:
is 5 randomly chosen?

Reply:
I've chosen to change `shared_client.go` as little as possible from the code in `cilium/dns` that's in use today, since minimal changes are encouraged. Hence please see https://github.com/cilium/dns/pull/13 for details on this code.